File: apps/api/services/finance_provisioning_dispatch.py
from __future__ import annotations

import asyncio
from contextlib import suppress
from dataclasses import dataclass

import httpx
from sqlalchemy import or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from core.config import Settings, get_settings
from core.logging import get_logger
from core.timestamps import now_unix_seconds
from db.models import FinanceProvisioningOutbox, ProvisioningRun
from db.repositories.provisioning_runs import ProvisioningRunRepository
from db.session import AsyncSessionLocal
from services.finance_provisioning import finance_event_payload

logger = get_logger(__name__)

_LOCK_TIMEOUT_SECONDS = 5 * 60
_MAX_ERROR_LENGTH = 2_000


@dataclass(frozen=True)
class FinanceDispatchSummary:
    claimed: int
    delivered: int
    failed: int
    configured: bool


async def claim_finance_provisioning_events(
    db: AsyncSession,
    *,
    now: int,
    limit: int,
) -> list[FinanceProvisioningOutbox]:
    """Claim one retry-safe batch without holding locks during network I/O."""
    retryable = or_(
        FinanceProvisioningOutbox.status.in_(("pending", "failed")),
        (
            (FinanceProvisioningOutbox.status == "processing")
            & (FinanceProvisioningOutbox.locked_at <= now - _LOCK_TIMEOUT_SECONDS)
        ),
    )
    rows = list(
        (
            await db.scalars(
                select(FinanceProvisioningOutbox)
                .where(
                    retryable,
                    FinanceProvisioningOutbox.available_at <= now,
                )
                .order_by(
                    FinanceProvisioningOutbox.created_at,
                    FinanceProvisioningOutbox.aggregate_id,
                    FinanceProvisioningOutbox.lifecycle_version,
                )
                .with_for_update(skip_locked=True)
                .limit(limit)
            )
        ).all()
    )
    run_ids = {run_id for row in rows if (run_id := getattr(row, "run_id", None)) is not None}
    runs_by_id: dict[str, ProvisioningRun] = {}
    if run_ids:
        runs = (
            await db.scalars(
                select(ProvisioningRun)
                .where(ProvisioningRun.id.in_(run_ids))
                .options(selectinload(ProvisioningRun.steps))
                .with_for_update()
            )
        ).all()
        runs_by_id = {run.id: run for run in runs}

    for row in rows:
        row.status = "processing"
        row.attempt_count += 1
        row.locked_at = now
        row.last_error = None
        row.updated_at = now
        run_id = getattr(row, "run_id", None)
        if run_id is not None and (run := runs_by_id.get(run_id)) is not None:
            ProvisioningRunRepository.mark_processing(run, now=now)
    await db.flush()
    return rows


async def dispatch_finance_provisioning_once(
    settings: Settings | None = None,
) -> FinanceDispatchSummary:
    active_settings = settings or get_settings()
    async with AsyncSessionLocal() as session:
        await ProvisioningRunRepository(session).expire_stale_application_runs(
            now=now_unix_seconds(),
            timeout_seconds=_LOCK_TIMEOUT_SECONDS,
        )
        await session.commit()
    billing_url = active_settings.billing_url.rstrip("/")
    if not billing_url or not active_settings.billing_internal_key:
        return FinanceDispatchSummary(claimed=0, delivered=0, failed=0, configured=False)

    async with AsyncSessionLocal() as session:
        rows = await claim_finance_provisioning_events(
            session,
            now=now_unix_seconds(),
            limit=active_settings.finance_provisioning_batch_size,
        )
        snapshots = [(row.id, row.attempt_count, finance_event_payload(row)) for row in rows]
        await session.commit()

    if not snapshots:
        return FinanceDispatchSummary(claimed=0, delivered=0, failed=0, configured=True)

    delivered = 0
    failed = 0
    endpoint = f"{billing_url}/api/v1/admin/finance-connections/ensure"
    async with httpx.AsyncClient(timeout=httpx.Timeout(15.0)) as client:
        for event_id, attempt_count, payload in snapshots:
            try:
                response = await client.post(
                    endpoint,
                    headers={
                        "x-internal-key": active_settings.billing_internal_key,
                        "content-type": "application/json",
                        "x-request-id": event_id,
                    },
                    json=payload,
                )
                response.raise_for_status()
            except Exception as exc:
                failed += 1
                await _mark_failed(event_id, attempt_count, _delivery_error(exc))
                logger.warning(
                    "finance_provisioning.delivery_failed",
                    event_id=event_id,
                    attempt_count=attempt_count,
                    error=_delivery_error(exc),
                )
            else:
                delivered += 1
                await _mark_delivered(event_id)
                logger.info(
                    "finance_provisioning.delivered",
                    event_id=event_id,
                    attempt_count=attempt_count,
                )

    return FinanceDispatchSummary(
        claimed=len(snapshots),
        delivered=delivered,
        failed=failed,
        configured=True,
    )


async def run_finance_provisioning_worker(
    stop: asyncio.Event,
    settings: Settings | None = None,
) -> None:
    """Continuously drain the durable outbox until application shutdown."""
    active_settings = settings or get_settings()
    while not stop.is_set():
        try:
            await dispatch_finance_provisioning_once(active_settings)
        except Exception:
            logger.error("finance_provisioning.worker_failed", exc_info=True)

        with suppress(TimeoutError):
            await asyncio.wait_for(
                stop.wait(),
                timeout=active_settings.finance_provisioning_poll_seconds,
            )


async def _mark_delivered(event_id: str) -> None:
    async with AsyncSessionLocal() as session:
        event = await session.get(FinanceProvisioningOutbox, event_id, with_for_update=True)
        if event is None or event.status != "processing":
            return
        now = now_unix_seconds()
        event.status = "delivered"
        event.delivered_at = now
        event.locked_at = None
        event.last_error = None
        event.updated_at = now
        if event.run_id:
            run = await ProvisioningRunRepository(session).retrieve(event.run_id, for_update=True)
            if run is not None:
                ProvisioningRunRepository.mark_succeeded(run, now=now)
        await session.commit()


async def _mark_failed(event_id: str, attempt_count: int, message: str) -> None:
    async with AsyncSessionLocal() as session:
        event = await session.get(FinanceProvisioningOutbox, event_id, with_for_update=True)
        if event is None or event.status != "processing":
            return
        now = now_unix_seconds()
        retry_seconds = min(3_600, 5 * (2 ** min(attempt_count, 10)))
        event.status = "failed"
        available_at = now + retry_seconds
        event.available_at = available_at
        event.locked_at = None
        event.last_error = message[:_MAX_ERROR_LENGTH]
        event.updated_at = now
        if event.run_id:
            run = await ProvisioningRunRepository(session).retrieve(event.run_id, for_update=True)
            if run is not None:
                ProvisioningRunRepository.mark_failed(
                    run,
                    now=now,
                    available_at=available_at,
                    message=event.last_error,
                )
        await session.commit()


def _delivery_error(error: Exception) -> str:
    if isinstance(error, httpx.HTTPStatusError):
        body = error.response.text[:500].strip()
        return f"Billing returned HTTP {error.response.status_code}: {body}"
    return f"{type(error).__name__}: {error}"

File: apps/api/src/workers/finance-provisioning-dispatch.ts
import {
  FinanceProvisioningContractError,
  parseFinanceProvisioningEvent,
  parseFinanceProvisioningReceipt,
  type FinanceProvisioningReceipt,
} from '@876/server/finance-provisioning'

import { getSettings } from '@/config'
import { AppHttpError } from '@/http/errors'
import { getLogger } from '@/platform/logger'
import { nowUnixSeconds } from '@/platform/timestamps'
import { financeEventPayload } from '@/services/finance-provisioning'

import {
  claimFinanceProvisioningEvents,
  claimFinanceProvisioningEventsByIds,
  expireStaleApplicationRuns,
  getFinanceProvisioningEventsByIds,
  markFinanceProvisioningDelivered,
  markFinanceProvisioningFailed,
} from './finance-provisioning-dispatch.repository'

const logger = getLogger('finance-provisioning-dispatch')

export type FinanceDispatchSummary = {
  claimed: number
  delivered: number
  failed: number
  configured: boolean
}

export type FinanceEnsureResult = FinanceDispatchSummary & {
  ensured: number
}

function deliveryError(error: unknown): string {
  if (error instanceof FinanceProvisioningContractError) {
    return `Finance provisioning contract ${error.issue}: ${error.message}`.slice(
      0,
      2000
    )
  }
  if (
    error instanceof Error &&
    error.message.startsWith('Billing returned HTTP')
  ) {
    return error.message.slice(0, 2000)
  }
  if (error instanceof Error) {
    return `${error.name}: ${error.message}`.slice(0, 2000)
  }
  return `${String(error)}`.slice(0, 2000)
}

async function postWithTimeout(
  url: string,
  headers: Record<string, string>,
  body: unknown,
  timeoutMs: number
): Promise<Response> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeoutMs)
  try {
    return await fetch(url, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
      signal: controller.signal,
    })
  } finally {
    clearTimeout(timer)
  }
}

export type FinanceDeliveryRow = {
  id: string
  attemptCount: number
  payload: Record<string, unknown>
}

/**
 * Performs the API → Billing wire hop without mutating outbox state.
 *
 * A 2xx status alone is not delivery. The response must satisfy the shared
 * finance receipt contract and prove a lifecycle at least as new as the event.
 * Foreground readiness additionally requires the exact lifecycle/status that
 * activation requested, preventing a stale or superseding state from being
 * mistaken for a ready workspace.
 */
export async function deliverFinanceProvisioningEvent(options: {
  eventId: string
  payload: unknown
  billingUrl: string
  internalKey: string
  exactState?: boolean
  timeoutMs?: number
}): Promise<FinanceProvisioningReceipt> {
  const event = parseFinanceProvisioningEvent(options.payload)
  if (event.eventId !== options.eventId) {
    throw new FinanceProvisioningContractError(
      'invalid-event',
      `Outbox event id ${options.eventId} does not match payload event id ${event.eventId}.`
    )
  }

  const baseUrl = options.billingUrl.trim().replace(/\/+$/, '')
  const endpoint = `${baseUrl}/api/v1/admin/finance-connections/ensure`
  const response = await postWithTimeout(
    endpoint,
    {
      'x-internal-key': options.internalKey,
      'content-type': 'application/json',
      'x-request-id': options.eventId,
    },
    event,
    options.timeoutMs ?? 15_000
  )

  if (!response.ok) {
    const text = await response.text()
    const snippet = text.slice(0, 500).trim()
    throw new Error(`Billing returned HTTP ${response.status}: ${snippet}`)
  }

  let body: unknown
  try {
    body = await response.json()
  } catch {
    throw new FinanceProvisioningContractError(
      'invalid-response',
      `Billing returned HTTP ${response.status} with a non-JSON success body.`
    )
  }

  return parseFinanceProvisioningReceipt(event, body, {
    exactState: options.exactState,
  })
}

async function deliverClaimedRows(
  rows: FinanceDeliveryRow[],
  billingUrl: string,
  internalKey: string,
  options: { exactState?: boolean } = {}
): Promise<{ delivered: number; failed: number }> {
  let delivered = 0
  let failed = 0
  for (const item of rows) {
    try {
      const receipt = await deliverFinanceProvisioningEvent({
        eventId: item.id,
        payload: item.payload,
        billingUrl,
        internalKey,
        exactState: options.exactState,
      })
      delivered += 1
      await markFinanceProvisioningDelivered(item.id, nowUnixSeconds())
      logger.info(
        {
          event_id: item.id,
          attempt_count: item.attemptCount,
          tenant_id: receipt.tenantId,
          connection_status: receipt.status,
          lifecycle_version: receipt.lifecycleVersion,
          applied: receipt.applied,
          duplicate: receipt.duplicate,
        },
        'finance_provisioning.delivered'
      )
    } catch (error) {
      failed += 1
      const message = deliveryError(error)
      await markFinanceProvisioningFailed(
        item.id,
        item.attemptCount,
        message,
        nowUnixSeconds()
      )
      logger.warn(
        { event_id: item.id, attempt_count: item.attemptCount, error: message },
        'finance_provisioning.delivery_failed'
      )
    }
  }
  return { delivered, failed }
}

export async function dispatchFinanceProvisioningOnce(): Promise<FinanceDispatchSummary> {
  // Expire stale application provisioning runs before claiming — matches Python
  // `dispatch_finance_provisioning_once` which runs this in its own transaction.
  try {
    await expireStaleApplicationRuns(nowUnixSeconds(), 5 * 60)
  } catch (error) {
    logger.error({ err: error }, 'finance_provisioning.expire_stale_failed')
  }

  const settings = getSettings()
  const billingUrl = settings.billing.url.trim().replace(/\/+$/, '')
  const internalKey = settings.billing.internalKey.trim()

  if (!billingUrl || !internalKey) {
    logger.error(
      {
        has_billing_url: Boolean(billingUrl),
        has_internal_key: Boolean(internalKey),
      },
      'finance_provisioning.not_configured'
    )
    return { claimed: 0, delivered: 0, failed: 0, configured: false }
  }

  const now = nowUnixSeconds()
  const limit = settings.billing.financeProvisioningBatchSize
  const claimedRows = await claimFinanceProvisioningEvents(now, limit)

  if (claimedRows.length === 0) {
    return { claimed: 0, delivered: 0, failed: 0, configured: true }
  }

  const snapshots: FinanceDeliveryRow[] = claimedRows.map((row) => ({
    id: row.id,
    attemptCount: row.attemptCount,
    payload: financeEventPayload(row as never),
  }))

  const { delivered, failed } = await deliverClaimedRows(
    snapshots,
    billingUrl,
    internalKey
  )

  return { claimed: snapshots.length, delivered, failed, configured: true }
}

export async function ensureFinanceProvisioningDelivered(
  eventIds: string[]
): Promise<FinanceEnsureResult> {
  const uniqueIds = [...new Set(eventIds)]
  if (uniqueIds.length === 0) {
    return { claimed: 0, delivered: 0, failed: 0, configured: true, ensured: 0 }
  }
  const settings = getSettings()
  const billingUrl = settings.billing.url.trim().replace(/\/+$/, '')
  const internalKey = settings.billing.internalKey.trim()
  if (!billingUrl || !internalKey) {
    logger.error(
      {
        has_billing_url: Boolean(billingUrl),
        has_internal_key: Boolean(internalKey),
        event_ids: uniqueIds,
      },
      'finance_provisioning.targeted_not_ready'
    )
    throw new AppHttpError({
      code: 'provisioning/finance-workspace-unavailable',
      message:
        'The finance workspace could not be prepared. Billing is not configured.',
      httpStatus: 503,
    })
  }
  logger.info({ event_ids: uniqueIds }, 'finance_provisioning.targeted_started')
  const now = nowUnixSeconds()
  const claimedRows = await claimFinanceProvisioningEventsByIds(now, uniqueIds)
  const snapshots: FinanceDeliveryRow[] = claimedRows.map((row) => ({
    id: row.id,
    attemptCount: row.attemptCount,
    payload: financeEventPayload(row as never),
  }))
  let delivered = 0
  let failed = 0
  if (snapshots.length > 0) {
    const result = await deliverClaimedRows(
      snapshots,
      billingUrl,
      internalKey,
      {
        exactState: true,
      }
    )
    delivered = result.delivered
    failed = result.failed
  }
  const states = await getFinanceProvisioningEventsByIds(uniqueIds)
  if (states.length !== uniqueIds.length) {
    const found = new Set(states.map((s) => s.id))
    const missing = uniqueIds.filter((id) => !found.has(id))
    logger.error(
      { missing_ids: missing, event_ids: uniqueIds },
      'finance_provisioning.targeted_not_ready'
    )
    throw new AppHttpError({
      code: 'provisioning/finance-workspace-unavailable',
      message:
        'The finance workspace could not be prepared. Missing provisioning event.',
      httpStatus: 503,
    })
  }
  const notDelivered = states.filter((s) => s.status !== 'delivered')
  if (notDelivered.length > 0) {
    for (const ev of notDelivered) {
      logger.error(
        {
          event_id: ev.id,
          organization_id: ev.organizationId,
          source_app_id: ev.sourceAppId,
          run_id: ev.runId,
          status: ev.status,
          attempt_count: ev.attemptCount,
          available_at: String(ev.availableAt),
          last_error: ev.lastError,
        },
        'finance_provisioning.targeted_not_ready'
      )
    }
    throw new AppHttpError({
      code: 'provisioning/finance-workspace-unavailable',
      message: 'The finance workspace could not be prepared. Please retry.',
      httpStatus: 503,
    })
  }
  for (const ev of states) {
    logger.info(
      {
        event_id: ev.id,
        organization_id: ev.organizationId,
        source_app_id: ev.sourceAppId,
        run_id: ev.runId,
        attempt_count: ev.attemptCount,
        status: ev.status,
      },
      'finance_provisioning.targeted_delivered'
    )
  }
  return {
    claimed: claimedRows.length,
    delivered,
    failed,
    configured: true,
    ensured: states.length,
  }
}

function sleepWithAbort(ms: number, signal?: AbortSignal): Promise<void> {
  if (!signal) return new Promise((resolve) => setTimeout(resolve, ms))
  if (signal.aborted) return Promise.resolve()
  return new Promise<void>((resolve) => {
    const timeout = setTimeout(() => {
      signal.removeEventListener('abort', onAbort)
      resolve()
    }, ms)
    const onAbort = (): void => {
      clearTimeout(timeout)
      signal.removeEventListener('abort', onAbort)
      resolve()
    }
    signal.addEventListener('abort', onAbort, { once: true })
  })
}

export async function runFinanceProvisioningWorker(options?: {
  signal?: AbortSignal
}): Promise<void> {
  const settings = getSettings()

  while (!options?.signal?.aborted) {
    try {
      await dispatchFinanceProvisioningOnce()
    } catch (error) {
      logger.error({ err: error }, 'finance_provisioning.worker_failed')
    }

    if (options?.signal?.aborted) break
    await sleepWithAbort(
      settings.billing.financeProvisioningPollSeconds * 1000,
      options?.signal
    )
  }
}

export function startFinanceProvisioningWorker(options?: {
  signal?: AbortSignal
}): { stop: () => Promise<void> } {
  const controller = new AbortController()
  const externalSignal = options?.signal
  if (externalSignal) {
    if (externalSignal.aborted) controller.abort()
    else
      externalSignal.addEventListener('abort', () => controller.abort(), {
        once: true,
      })
  }

  const done = runFinanceProvisioningWorker({ signal: controller.signal })
  void done.catch((err) =>
    logger.error({ err }, 'finance_provisioning.worker_failed')
  )

  return {
    stop: async () => {
      controller.abort()
      try {
        await done
      } catch {}
    },
  }
}
